lua/lspconfig/vdmj.lua

```diff
@@ -0,0 +1,99 @@
+local configs = require 'lspconfig/configs'
+local util = require 'lspconfig/util'
+
+local server_name = 'vdmj'
+
+local function getJarPath(config, package, version)
+  return util.path.join(config.options.mavenrepo, package, version, package .. '-' .. version .. '.jar')
+end
+
+local function get_version(config)
+  if config.options.high_precision == true then
+    return config.options.version:gsub('([%d.]+)', '%1-P')
+  else
+    return config.options.version
+  end
+end
+
+configs[server_name] = {
+  default_config = {
+    cmd = { 'java' },
+    filetypes = { 'vdm' },
+    root_dir = function(fname)
+      return util.find_git_ancestor(fname) or vim.fn.getcwd()
+    end,
+    options = {
+      java = tostring(vim.fn.getenv 'JAVA_HOME') or '/usr' .. '/bin/java',
+      java_opts = { '-Xmx3000m', '-Xss1m' },
+      mavenrepo = util.path.join(tostring(vim.fn.getenv 'HOME'), '/.m2/repository/com/fujitsu'),
+      logfile = util.path.join(vim.fn.stdpath 'cache', 'vdm-lsp.log'),
+      debugger_port = 0,
+      high_precision = false,
+    },
+  },
+  docs = {
+    description = [[
+https://github.com/nickbattle/vdmj
+
+The VDMJ language server can be installed by cloning the VDMJ repository and
+running `mvn clean install`.
+
+Various options are provided to configure the language server, with the
+following being mandatory:
+
+- `version`,
+- `dialect` (one of `vdmsl`, `vdmpp` or `vdmrt`).
+
+```lua
+require'lspconfig'.vdmj.setup {
+  options = {
+    version = '4.4.3-SNAPSHOT',
+    dialect = 'vdmsl',
+    ...
+  }
+  ...
+}
+```
+
+A value of `0` for `debugger_port` will disable the debugger. Note that if a
+non-zero value is used, only one instance of the server can be active at a
+time.
+
+Note: proof obligations and combinatorial testing are not currently supported
+by neovim.
+]],
+    default_config = {
+      cmd = "Generated from the options given",
+      root_dir = "Git root or vim's starting directory",
+      options = {
+        java = '$JAVA_HOME/bin/java',
+        mavenrepo = '$HOME/.m2/repository/com/fujitsu',
+        logfile = "path.join(vim.fn.stdpath 'cache', 'vdm-lsp.log')",
+        debugger_port = 0,
+        high_precision = false,
+      },
+    },
+  },
+  on_new_config = function(config)
+    local version = get_version(config)
+    config.cmd = {
+      config.options.java,
+      -- TODO: how to unpack java_opts here? unpack(config.options.java_opts)
+      -- results in only the first option being used.
+      '-Xmx3000m',
+      '-Xss1m',
+      '-Dlsp.log.filename=' .. config.options.logfile,
+      '-cp',
+      table.concat({
+        getJarPath(config, 'vdmj', version),
+        getJarPath(config, 'annotations', version),
+        getJarPath(config, 'lsp', version),
+      }, ':'),
+      'lsp.LSPServerStdio',
+      '-' .. config.options.dialect,
+      '-dap',
+      -- TODO: should be unique for each instance
+      tostring(config.options.debugger_port),
```

Comment:
A new server will fail to start if the port is already in use. Is there a way to increment the port number either for each new configuration or if the port is already in use?

Reply:
Usually binding to 0 picks a random available port. Not sure if that's the case here though. Otherwise I think the only way is to attempt to bind to a port (see `:h tcp-server` for examples).

Is it necessary/important to bind to a port at all?

Reply:
> Usually binding to 0 picks a random available port. Not sure if that's the case here though.

It is not the case at the moment with the VDMJ LS, unfortunately.

> Is it necessary/important to bind to a port at all?

Binding a port is only necessary to start the debugger adapter with the LS. It is not necessary if the DA is not needed.

Reply:
> Usually binding to 0 picks a random available port.

Based on this, I changed the configuration to disable the debugger for negative port values. I don't know whether it is better to have the logic for using a random port in the configuration or in the LS itself, so I think it is best to leave it out for the moment.